### PR TITLE
PMEM allocation 

### PR DESCRIPTION
**Pull request by ankit-iitb**

### CODE REVIEW

File: kernel/run.py [hunk 1]
```diff
@@ -372,23 +382,19 @@ def pmem_paths(args):
...
                     pmem = "on"
                 else:
-                    if len(paths) > 0:
-                        paths.append(paths[node % num_host_numa_nodes])
-                    else:
-                        f = open("{}".format(pmem_test_path), "w+")
-                        paths += [pmem_test_path] * args.qemu_nodes
-                        return paths
+                    path = "{}-{}".format(pmem_test_path, node)
+                    open(path, "w+")
```
Comment (gz) [inline kernel/run.py:386]: what does this line do? return result is not stored anywhere?

File: kernel/run.py [hunk 2]
```diff
@@ -372,23 +382,19 @@ def pmem_paths(args):
...
+                    paths.append(path)
         return paths
 
     host_numa_nodes_list = query_host_numa()
     num_host_numa_nodes = len(host_numa_nodes_list)
     if args.qemu_nodes and args.qemu_nodes > 0 and args.qemu_cores > 1:
-        if int(args.qemu_pmem):
-            pmem_paths = pmem_paths(args)
-            assert len(pmem_paths) == args.qemu_nodes
+        if args.qemu_pmem and int(args.qemu_pmem):
```
Comment (gz) [inline kernel/run.py:393]: can we avoid this if we do `type=int` in the parser?

should probably do the same for qemu-memory then?
```
parser.add_argument("--qemu-memory", type=int,
                    help="How much total memory in MiB (will get evenly divided among nodes).", default=1024)
parser.add_argument("--qemu-pmem", type=int,
                    help="How much total peristent memory in MiB (will get evenly divided among nodes).", required=False, default=None)
```

File: kernel/run.py [hunk 3]
```diff
@@ -402,13 +408,14 @@ def pmem_paths(args):
...
+            if args.qemu_pmem and int(args.qemu_pmem):
+                pmem_per_node = int(args.qemu_pmem) / args.qemu_nodes
                 qemu_default_args += ['-object', 'memory-backend-file,id=pmem{},mem-path={},size={}M,pmem={},share=on'.format(
-                    node, pmem_paths[node], int(pmem_per_node), pmem)]
+                    node, pm_paths[node], int(pmem_per_node), pmem)]
                 qemu_default_args += ['-device',
                                     'nvdimm,node={},slot={},id=nvdimm{},memdev=pmem{}'.format(node, node, node, node)]
 
-    if args.qemu_nodes and args.qemu_nodes > 0 and int(args.qemu_pmem):
+    if args.qemu_pmem and int(args.qemu_pmem):
```
Comment (gz) [inline kernel/run.py:418]: type=int? see above

File: kernel/run.py [hunk 4]
```diff
@@ -402,13 +408,14 @@ def pmem_paths(args):
             qemu_default_args += ["-numa", "cpu,node-id={},socket-id={}".format(
                 node, node)]
             # NVDIMM related arguments
-            if int(args.qemu_pmem):
```
Comment (gz) [inline kernel/run.py:405]: type=int see above?

File: kernel/run.py [hunk 5]
```diff
@@ -418,7 +425,7 @@ def pmem_paths(args):
                               "{},sockets=1".format(args.qemu_cores)]
 
     if args.qemu_memory:
-        if args.qemu_nodes and int(args.qemu_pmem):
+        if args.qemu_pmem and int(args.qemu_pmem):
```
Comment (gz) [inline kernel/run.py:428]: type=int?

File: kernel/run.py [hunk 6]
```diff
@@ -504,8 +511,9 @@ def pmem_paths(args):
             print("STDERR: {}".format(execution.stderr.decode('utf-8')))
 
     # If the test creates a fake pmem path; remove it.
-    if os.path.isfile(pmem_test_path):
-        os.remove(pmem_test_path)
+    for path in pmem_paths(args):
```
Comment (gz) [inline kernel/run.py:514]: can we use `with` block and [NamedTemporaryFile](https://docs.python.org/2/library/tempfile.html#tempfile.NamedTemporaryFile) so this gets cleaned up even if something fails randomly further up?

File: kernel/src/arch/x86_64/mod.rs [hunk 7]
```diff
@@ -475,6 +506,54 @@ fn identify_numa_affinity(
     );
 }
 
+/// Map the persistent memory addresses to the vspace.
+fn map_physical_persistent_memory() {
```
Comment (gz) [inline kernel/src/arch/x86_64/mod.rs:510]: is there a reason we can't do this as part of the bootloader?

File: kernel/src/arch/x86_64/mod.rs [hunk 8]
```diff
@@ -689,6 +768,49 @@ fn _start(argc: isize, _argv: *const *const u8) -> isize {
         kcb.set_physical_memory_manager(tcache);
     }
 
+    // 1. Discover persistent memory using topology information.
```
Comment (gz) [inline kernel/src/arch/x86_64/mod.rs:771]: e.g., same question as above might be better if we can somehow make it part of bootlaoder?

File: kernel/src/arch/x86_64/syscall.rs [hunk 9]
```diff
@@ -258,10 +259,15 @@ fn handle_vspace(arg1: u64, arg2: u64, arg3: u64) -> Result<(u64, u64), KError>
     let mut p = kcb.arch.current_executor()?;
 
     match op {
-        VSpaceOperation::Map => unsafe {
+        VSpaceOperation::Map | VSpaceOperation::MapPM => unsafe {
```
Comment (gz) [inline kernel/src/arch/x86_64/syscall.rs:262]: maybe we should rename those enum variants to MapMem and MapPMem now?

File: kernel/src/arch/x86_64/syscall.rs [hunk 10]
```diff
@@ -335,7 +345,7 @@ fn handle_vspace(arg1: u64, arg2: u64, arg3: u64) -> Result<(u64, u64), KError>
             )?;
             Ok((paddr.as_u64(), size as u64))
         },
-        VSpaceOperation::Unmap => {
+        VSpaceOperation::Unmap | VSpaceOperation::UnmapPM => {
```
Comment (gz) [inline kernel/src/arch/x86_64/syscall.rs:348]: same UnmapMem, UnmapPMem?

File: kernel/src/kcb.rs [hunk 11]
```diff
@@ -355,6 +364,10 @@ impl<A: ArchSpecificKcb> Kcb<A> {
         self.physical_memory.gmanager = Some(gm);
     }
 
+    pub fn set_global_pmem(&mut self, gm: &'static GlobalMemory) {
+        self.pmem_memory.gmanager = Some(gm);
+    }
+
     pub fn set_allocation_affinity(&mut self, node: atopology::NodeId) -> Result<(), KError> {
```
Comment (gz) [inline kernel/src/kcb.rs:371]: set_mem_affinity for namign consistency?

File: kernel/src/kcb.rs [hunk 12]
```diff
@@ -383,10 +396,42 @@ impl<A: ArchSpecificKcb> Kcb<A> {
         }
     }
 
+    pub fn set_pmem_affinity(&mut self, node: atopology::NodeId) -> Result<(), KError> {
```
Comment (gz) [inline kernel/src/kcb.rs:399]: I also wonder can we make a generic set_allocation_affinity that gets the either pmem_arenas or memory_arenas as an argument and avoid code duplication here?

File: kernel/src/kcb.rs [hunk 13]
```diff
@@ -383,10 +396,42 @@ impl<A: ArchSpecificKcb> Kcb<A> {
...
+            core::mem::swap(&mut arena, &mut self.pmem_memory);
+            self.pmem_arenas[arena.affinity as usize].replace(arena);
+
+            Ok(())
+        } else {
+            Err(KError::InvalidAffinityId)
+        }
+    }
+
     pub fn set_physical_memory_manager(&mut self, pmanager: TCache) {
```
Comment (gz) [inline kernel/src/kcb.rs:427]: maybe rename to set_mem_manager for consistency

File: kernel/src/kcb.rs [hunk 14]
```diff
@@ -355,6 +364,10 @@ impl<A: ArchSpecificKcb> Kcb<A> {
         self.physical_memory.gmanager = Some(gm);
```
Comment (gz) [inline kernel/src/kcb.rs:364]: rename to `set_global_mem` for naming consistency (persistent memory -> pmem, volatile memory -> mem)

File: lib/kpi/src/lib.rs [hunk 15]
```diff
@@ -135,6 +135,17 @@ impl From<&str> for ProcessOperation {
     }
 }
 
+/// Flags to specify the memory type
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+#[repr(u64)]
+pub enum MemType {
+    /// The allocations will be done from DRAM.
+    DRAM = 1,
```
Comment (gz) [inline lib/kpi/src/lib.rs:143]: Mem?

File: lib/kpi/src/lib.rs [hunk 16]
```diff
@@ -135,6 +135,17 @@ impl From<&str> for ProcessOperation {
...
+    /// The allocations will be done from PMEM.
+    PMEM = 2,
```
Comment (gz) [inline lib/kpi/src/lib.rs:145]: PMem?

File: lib/kpi/src/lib.rs [hunk 17]
```diff
@@ -149,6 +160,10 @@ pub enum VSpaceOperation {
     MapFrame = 4,
```
Comment (gz) [inline lib/kpi/src/lib.rs:160]: MapMemFrame?

File: kernel/tests/integration-test.rs [hunk 18]
```diff
@@ -2282,3 +2282,29 @@ fn s06_leveldb_benchmark() {
         check_for_successful_exit(&cmdline, qemu_run(), output);
     }
 }
+
+/// Tests that basic pmem allocation support is functional.
+#[test]
+fn s06_pmem_alloc() {
+    let machine = Machine::determine();
```
Comment (gz) [inline kernel/tests/integration-test.rs:2289]: probably want to make sure that the test can persist data durably too eventually?

**Review by gz — CHANGES_REQUESTED** [review]

**Comment by gz** [comment]
Regarding the bootloader comment, do the mappings not show up as PERSISTENT_MEMORY [here](https://github.com/vmware-labs/node-replicated-kernel/blob/master/bootloader/src/main.rs#L193)?

File: kernel/run.py (hunk 1 above)
Comment (ankit-iitb) [reply kernel/run.py]: Thank you for the fantastic review. It was helpful.

I completely changed the run.py PMem related code. Not applicable anymore.

File: kernel/run.py (hunk 6 above)
Comment (ankit-iitb) [reply kernel/run.py]: Thanks, I didn't know that something like this existed. Updated; could you please check if the changes rectify your concerns.

File: kernel/src/arch/x86_64/mod.rs (hunk 7 above)
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/mod.rs:509]: As discussed earlier, UEFI does not handle PMem related tables. Bespin gets PMem related information through ACPI tables (using topology crate).

File: kernel/src/arch/x86_64/mod.rs (hunk 8 above)
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/mod.rs:770]: See above.

File: kernel/src/arch/x86_64/syscall.rs (hunk 9 above)
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/syscall.rs]: Done.

File: kernel/src/arch/x86_64/syscall.rs (hunk 10 above)
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/syscall.rs]: Updated.

File: kernel/src/kcb.rs (hunk 11 above)
Comment (ankit-iitb) [reply kernel/src/kcb.rs]: Updated.

File: kernel/src/kcb.rs (hunk 13 above)
Comment (ankit-iitb) [reply kernel/src/kcb.rs]: Updated.

File: kernel/src/kcb.rs (hunk 14 above)
Comment (ankit-iitb) [reply kernel/src/kcb.rs:364]: Updated.

File: lib/kpi/src/lib.rs (hunk 15 above)
Comment (ankit-iitb) [reply lib/kpi/src/lib.rs]: Updated.

File: lib/kpi/src/lib.rs (hunk 16 above)
Comment (ankit-iitb) [reply lib/kpi/src/lib.rs]: Updated.

File: lib/kpi/src/lib.rs (hunk 17 above)
Comment (ankit-iitb) [reply lib/kpi/src/lib.rs]: Updated.

File: kernel/tests/integration-test.rs (hunk 18 above)
Comment (ankit-iitb) [reply kernel/tests/integration-test.rs:2290]: Added a TODO statement.

File: kernel/src/kcb.rs (hunk 12 above)
Comment (ankit-iitb) [reply kernel/src/kcb.rs:399]: The pattern seems weird; the outer layer will be sending the information to `set_allocation_affinity` that is already a part of KCB (arena and mem manager).

If you want to make one copy, maybe I can add a macro and change the variable names such that the shared code would be in the macro?

File: kernel/run.py (hunk 2 above)
Comment (ankit-iitb) [reply kernel/run.py]: Done.

File: kernel/run.py (hunk 3 above)
Comment (ankit-iitb) [reply kernel/run.py]: Done.

File: kernel/run.py (hunk 4 above)
Comment (ankit-iitb) [reply kernel/run.py:405]: Done.

File: kernel/run.py (hunk 5 above)
Comment (ankit-iitb) [reply kernel/run.py]: Done.

File: kernel/src/arch/x86_64/mod.rs (hunk 7 above)
Comment (gz) [reply kernel/src/arch/x86_64/mod.rs:514]: Yea strange. FWIW I just tried to upgrade our OVMF to the latest nightly and the pmem regions still don't show up when we query the uef memory map in the bootloader :/

File: kernel/src/kcb.rs (hunk 12 above)
Comment (gz) [reply kernel/src/kcb.rs:415]: Committed something let me know if this is bad if I missed something

Comment (ankit-iitb) [reply kernel/src/kcb.rs:415]: Ah, I "somehow" assumed that you want one `pub fn set_allocation_affinity` that others call. Hence the comment about passing arenas from outside.

This makes more sense, separate functions for Mem and PMem, and shared logic internally.

File: kernel/src/arch/x86_64/mod.rs (hunk 7 above)
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/mod.rs:514]: We can keep the TODO comment. 

However, IIRC, SNIA intentionally didn't put the logic in UEFI but ACPI. They didn't want to change the firmware again and again for each change, and eventually, they also wanted to move towards a plug-and-play model for CXL(PCIe 5.0).

**Pull request merged**